### DISCUSSION TITLE
Hardcoded creature AI ranged attack cooldown?

### DISCUSSION THREAD

**Discussion by Ek-Videogames** [FakeFishGames/Barotrauma#8343] [Code questions and modding support]
I recently tried to convert some modded creatures to utilize the new proper ranged attack system instead of a hacky ranged/melee workaround, their attacks and aim animations work perfectly when controlled manually by a player in testing - however when their AI starts attacking trying to attack an enemy they just stand around and only fire a projectile every few seconds ( despite their limb's attack cooldown being much less - closer to 1 second )

I've noticed the behavior of ruin guardians often only fire their weapons intermittently as well, despite their cooldown defined in XML, so I just wanted to ask if this was just a quirk / intentional AI limitation?

If its possible to make the AI shoot more than it currently does, then what numbers do I need to make lower? 
Because I've already tried:
`cooldown, secondarycooldown, afterattackdelay, afterattack behavior type, react & attack distance, timer`

**Comment by itchyOwl** [comment]
There's no hard-coded cooldowns, but the system can be a bit confusing because it's built on top of an existing attack system based on melee attacks. (Also, defining things some things in the creature's AI element and other things in the attack elements in the ragdoll file is not optimal, but it's something we can't change anymore.)

I'm not sure what you mean with "the new proper ranged attack system" contra "a hacky ranged/melee workaround", but ranged attacks are actually a bit hacky: the creatures attack from a distance and also "hit" (or rather in this case trigger/launch the attack) from there, which is why you should set both `range` and `damagerange` to the same value (maybe we should always use the same values for ranged attacks though).

`cooldown` defines how often the creature attempts to attack, and `afterattack` how the creature moves under the cooldown period. `secondarycooldown` allows the creature to attack with other valid attacks even though the cooldown is active. E.g. Spinelings have multiple spikes that each are hidden after the attack for 15 seconds, but with a secondary cooldown of 0.3, so they can shoot _other_ spikes rapidly after the first. The giant spinelings shoots less frequently (3 sec). <s>The cooldown values don't actually have any relevance here, because the hidden spikes can't be shot.</s> Edit: Incorrect information: hidden limbs can be used for attacking.

With the default 0 value the creature cannot attack before the cooldown is expired. So to make the creature shoot more, just decrease the cooldown, e.g. Fractal Guardian 2 (with the steam gun) has 0 that allows it to shoot continuously.

`afterattackdelay` defines a delay after the attack and between the cooldown where the creature continues to move according to the attack pattern (e.g. towards the target) and before it starts to follow the movement behavior defined in `afterattack`. E.g. they creature doesn't turn and start falling back _immediatedly_ after hitting the target, but instead after a brief delay.

Other things to take into account:
- Attack's `duration` defines how long the creature attacks, so if this value is too low, the hit doesn't register.
- Ranged attacks also have the `requiredangle` parameter, which limits when the attack is considered valid: the limb has to point approximately to the target before the creature can choose the attack. If the attack shouldn't be affected by this limitation, use `requiredangle="360"`. For example, if the creature uses hands to hold a weapon and actually aims at the target, like Fractal Guardians.

The guardians use items for shooting, so the cooldown defined in the item also has an effect. The one with the harpoon shoots only once and the other with the steamgun should shoot continuously. If you find that it's not working like that, please post a bug report about it.

So to answer the question in short: just test lowering the cooldown of Crawler Broodmother's ranged attack to 1 and see how that works (You might want to remove the other attack definitions: "Water Attack C" and "Structure Attack").

Hope that this helps.

**Comment by itchyOwl** [comment]
Actually there is a certain hard coded cooldown value that affects ranged attacks: when the selected limb doesn't point to the target, we'll add some aim delay to prevent the creature from shooting immediatedly when it "spots" the target.

Found out that when the creature changed the attacking limb, we always triggered this cooldown. Changed it so that we'll now use the attack cooldown instead if it's lower than the hard coded default of 1.5s. If the cooldown was 0, this wasn't and isn't applied.

Fixed in Regalis11/Barotrauma-development@4f9c529.
